enhancements/network/ocp-on-smart-nics.md

```diff
@@ -0,0 +1,313 @@
+need to be kept in sync with the software on the host PC (eg, as new
+functionality is added to OCP). The obvious answer to this problem is
+to run OpenShift on the NIC as well as on the host.
+
+The initial target of this work is the [NVIDIA BlueField-2 DPU].
+However, similar products are being worked on by other vendors which
+we will eventually want to support, so the solution should not be too
+specialized to this one product.
+
+[NVIDIA BlueField-2 DPU]: https://www.nvidia.com/en-us/networking/products/data-processing-unit/
+
+## Motivation
+
+### Goals
+
+- Design a system for provisioning and maintaining a RHEL-based
+  software stack on Smart NICs to support features such as "[Smart NIC
+  OVN Offload]".
+
+    - Although this enhancement is currently written to assume that
+      this means running OCP on the NICs, this is not yet cast in
+      stone. Some other approaches are discussed under "Alternatives"
+      and one of those could potentially replace the OCP approach.
+
+    - Although this enhancement technically only covers the
+      installing/maintaining part of the problem, there will be heavy
+      overlap with the "Smart NIC OVN Offload" enhancement (with the
+      decisions that are made here affecting how OVN and
+      ovn-kubernetes get deployed there).
+
+    - Although not part of the initial design, we assume some
+      customers may be interested taking advantage of the NIC/host
+      separation to run additional security and monitoring
+      functionality on the NICs in the future. Thus, we are interested
+      in designs with more isolation between the NIC and the host,
+      even though that is not necessarily interesting for the primary
+      network-offload use case.
+
+- Design an appropriate architecture for an environment containing a
+  mix of x86_64 hosts with "dumb" NICs, x86_64 hosts with Smart NICs,
+  and the ARM-based Smart NICs themselves. (As discussed below, such
+  an environment may involve one or more OCP clusters.)
+
+- Define how OCP installs and upgrades will work in such an
+  environment.
+
+[Smart NIC OVN Offload]: https://github.com/openshift/enhancements/pull/732
+
+### Non-Goals
+
+- Support for network plugins other than OVN Kubernetes. The
+  architecture designed here may not be generically applicable and is
+  not guaranteed to be usable by other network plugins even if they
+  independently implement their own BlueField, etc, support.
+
+### Supported Hardware
+
+We want to support Smart NICs that are vaguely "BlueField-2-ish". eg:
+
+  1. The NIC has a sufficiently-powerful ARM processor (or,
+     theoretically, a processor of some other architecture that RHCOS
+     supports), and sufficient RAM, SSD, etc, to run as an OCP node.
+
+       - Current BlueField-2 specs mention "Up to 8 Armv8 A72 cores"
+         and "8GB / 16GB / 32GB of on-board DDR4". It is possible we
+         would only support a subset of SKUs.
```

Comment:
When I look the BF2 spec, BF2 seems not to have storage (e.g. ssd/hdd) hence BF2 should run kernel with ramdisk. This means BF2's RAM will be consumed not only by process working memory also by ramdisk for RHCOS/container images/logging data from container).

Currently OSP does not have strict storage management as far as I know. Is there any idea for that?

Reply:
It's got 32G of eMMC (ie, flash-card-on-a-chip), which is supposed to be divided between two installs (so, eg, you're running from partition 0 and install an upgrade into partition 1, then reboot into partition 1, and if the watchdog timer goes off before the card comes back up, it will automatically reboot back into partition 0 for you).

The [OCP requirements for bare metal](https://docs.openshift.com/container-platform/4.6/installing/installing_bare_metal/installing-bare-metal.html#minimum-resource-requirements_installing-bare-metal) say workers need at least 8G of storage, so we should be OK.

(The cards also have M.2 NVMe ports so you can attach SSDs to them.)

Reply:
At least 4.7, worker node requires 120G...
4.6 doc (you mentioined) seems broken (vCPU 2 goes RAM field).

https://docs.openshift.com/container-platform/4.7/installing/installing_bare_metal/installing-bare-metal.html#minimum-resource-requirements_installing-bare-metal

Reply:
Ah... well, immediately post-install, a bare-metal worker is using 8G of its disk (masters use a bit more), so the "120G required" seems to be more about assuming how much space the customer is _eventually_ going to need for their own workloads.

Still, we aren't going to have _lots_ of extra space (especially if we stick with two partitions).

Reply:
I do not only mention that initial required space also the space the process/container uses (at least, to keep the container image, cri-o consumes the space and many containers/processes uses the space for logging). 

I believe you also consider these stuff and conclude you can. I have concern so I just share it with you.

Note: please not add me into reviewer because I want not to be the one.